File: app/src/main/java/com/test/xyz/demo/presentation/projectlist/presenter/ProjectListPresenterImpl.java
package com.test.xyz.demo.presentation.projectlist.presenter;

import com.test.xyz.demo.R;
import com.test.xyz.demo.domain.interactor.project.ProjectInteractor;
import com.test.xyz.demo.domain.model.github.GitHubRepo;
import com.test.xyz.demo.presentation.common.DisposableManager;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;

public class ProjectListPresenterImpl implements ProjectListPresenter {
    private final ProjectListView projectListView;
    private final ProjectInteractor projectInteractor;
    private final DisposableManager disposableManager;

    @Inject
    public ProjectListPresenterImpl(ProjectListView projectListView, ProjectInteractor projectInteractor) {
        this.projectListView = projectListView;
        this.projectInteractor = projectInteractor;
        this.disposableManager = new DisposableManager();
    }

    @Override
    public void requestProjectList(String userName) {
        Disposable disposable = projectInteractor.getProjectList(userName)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableObserver<List<GitHubRepo>>() {
                    @Override
                    public void onNext(List<GitHubRepo> gitHubRepos) {
                        projectListView.showProjectList(gitHubRepos);
                    }

                    @Override
                    public void onError(Throwable e) {
                        projectListView.showError(R.string.project_list_ret_error);
                    }

                    @Override
                    public void onComplete() {
                    }
                });

        disposableManager.add(disposable);
    }

    @Override
    public void onStop() {
        disposableManager.dispose();
    }
}

File: app/src/main/kotlin/com/test/xyz/demo/presentation/projectlist/presenter/ProjectListPresenterImpl.kt
package com.test.xyz.demo.presentation.projectlist.presenter

import com.test.xyz.demo.R
import com.test.xyz.demo.domain.interactor.project.ProjectInteractor
import com.test.xyz.demo.domain.model.github.GitHubRepo
import com.test.xyz.demo.presentation.common.DisposableManager
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

class ProjectListPresenterImpl @Inject
constructor(private val projectListView: ProjectListView, private val projectInteractor: ProjectInteractor) : ProjectListPresenter {
    private val disposableManager: DisposableManager

    init {
        this.disposableManager = DisposableManager()
    }

    override fun requestProjectList(userName: String) {
        val disposable = projectInteractor.getProjectList(userName)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableObserver<List<GitHubRepo>>() {
                    override fun onNext(gitHubRepos: List<GitHubRepo>) {
                        projectListView.showProjectList(gitHubRepos)
                    }

                    override fun onError(e: Throwable) {
                        projectListView.showError(R.string.project_list_ret_error)
                    }

                    override fun onComplete() {}
                })

        disposableManager.add(disposable)
    }

    override fun onStop() {
        disposableManager.dispose()
    }
}
